appserver/connectors/connectors-runtime/src/main/java/com/sun/enterprise/resource/listener/LocalTxConnectionEventListener.java

```diff
@@ -72,12 +86,15 @@ public void connectionErrorOccurred(ConnectionEvent evt) {
      * @param evt ConnectionEvent
      */
     @Override
-    public void badConnectionClosed(ConnectionEvent evt) {
+    public synchronized void badConnectionClosed(ConnectionEvent evt) {
         Object connectionHandle = evt.getConnectionHandle();
         ResourceHandle handle = resource;
         if (associatedHandles.containsKey(connectionHandle)) {
```

Comment:
Here too

Reply:
Also, because `synchronized` is only used to guard access to the associated handles map, maybe it's better to use `ReentrantReadWriteLock` instead?